### PR TITLE
Feature/tracking tags

### PR DESCRIPTION
**Pull request by fgourichon**
@cesar-tonnoir for review

### CODE REVIEW

File: src/components/widgets-settings/tag-filter/tag-filter.directive.coffee [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+        # in the MVP, all rules are combined by OR, and all conditions within a rule a AND
+
+        filterQuery = {}
+        filterQuery['condition'] = 'OR'
+        filterQueryRules = []
+        filterQuery['rules'] = filterQueryRules
+
+        angular.forEach  scope.settingsTags, (settingsTag) ->
+          filterQueryCondition = {}
```
Comment (cesar-tonnoir) [inline src/components/widgets-settings/tag-filter/tag-filter.directive.coffee]: This is correct, but in coffeescript you can simplify the syntax a bit:
```coffeescript
filterQueryCondition = 
  condition: 'AND'
  rules: []
# => { 'condition' => 'AND', 'rules' => [] } 

ruleTag = 
  value: filterItem.value
ruleTag.name = filterItem.name if filterItem.name?
# => { 'value' => 'myvalue', 'name' => 'aname' } 
```

File: src/components/widgets-settings/tag-filter/tag-filter.directive.coffee [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        scope.settingsTags = filterToSettingTags (w.metadata.filter_query)
+
+      initiateAutoComplete= ->
+        tags = w.content.available_tags || []
+        autotags = []
+        for org, tag_hash of tags
+          angular.forEach tag_hash['tag_references'], (tag_ref) ->
+            angular.forEach tag_ref['tag_reference_values'], (tag) ->
+              # TODO: Put ruletag generation in a function
+              tagtext = if tag_ref['name']? then tag_ref['name'] + ":" + tag['value'] else tag['value']
```
Comment (cesar-tonnoir) [inline src/components/widgets-settings/tag-filter/tag-filter.directive.coffee]:
```coffeescript
tagtext = _.compact([tag_ref.name, tag_ref.value]).join(':')
```

File: src/components/widgets-settings/tag-filter/tag-filter.directive.coffee [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+              ruletag = {}
+              ruletag['text'] = tagtext
+              ruletag['name'] = tag_ref['name'] if tag_ref['name']?
+              ruletag['value'] = tag['value']
+              autotags.push(ruletag)
+
+        scope.loadTagList= (query) ->
+          return $.grep(autotags, (e) -> e.text.toLowerCase().indexOf(query.toLowerCase()) > -1)
```
Comment (cesar-tonnoir) [inline src/components/widgets-settings/tag-filter/tag-filter.directive.coffee]:
```coffeescript
return _.select(autotags, (e) -> e.text.toLowerCase().match(query.toLowerCase()))
```

File: src/components/widgets/accounts-balance-sheet/accounts-balance-sheet.directive.coffee [hunk 4]
```diff
@@ -28,6 +30,14 @@ module.controller('WidgetAccountsBalanceSheetCtrl', ($scope, $q, ImpacWidgetsSvc
   $scope.ascending = true
   $scope.sortedColumn = 'account'
 
+  # Display filter tags based on settings
+  # --------------------------------------
+  $scope.areFilterTagsEnabled = ->
+    ImpacTheming.get().widgetSettings.tagging.enableFilterTags
```
Comment (cesar-tonnoir) [inline src/components/widgets/accounts-balance-sheet/accounts-balance-sheet.directive.coffee]: Can you please rename to `widgetSettings.tagging.enabled`?

File: src/components/widgets/accounts-profit-and-loss/accounts-profit-and-loss.directive.coffee [hunk 5]
```diff
@@ -46,6 +48,14 @@ module.controller('WidgetAccountsProfitAndLossCtrl', ($scope, $q, ChartFormatter
       $scope.period_translation = _.capitalize(translation.toLowerCase())
   )
 
+  # Display filter tags based on settings
+  # --------------------------------------
+  $scope.areFilterTagsEnabled = ->
```
Comment (cesar-tonnoir) [inline src/components/widgets/accounts-profit-and-loss/accounts-profit-and-loss.directive.coffee]: just a detail - In this case you can move it to `.init()` and assign a variable `$scope.areFilterTagsEnabled` => if you inject a method in the template, it's going to be evaluated at each digest cycle

File: src/services/theming/theming.svc.coffee [hunk 6]
```diff
@@ -111,6 +111,15 @@ angular
             default: "impac.widget.theming.hist.default.prefix"
         timePeriod:
           showSlider: true
+        tagging:
+          enableFilterTags: true
+
+      translateSettings:
```
Comment (cesar-tonnoir) [inline src/services/theming/theming.svc.coffee]: merging problem?

File: DEVELOPER.md [hunk 7]
```diff
@@ -310,6 +310,15 @@ You want to make the generator better? Of course. See the [README](./generators/
 5. Notify the parent widget that the setting's context has been loaded and is ready: `$scope.deferred.resolve(setting)`.
 **IMPORTANT**: The parent widget's #show method (= call to the Impac! API to retrieve the widget's content) will be called only once all the settings are loaded (= once they have resolved their `$scope.deferred` object). 
 
+6. Add the new components angular module to the `src/impac-angular.module.js` module declarations.
```
Comment (fgourichon) [inline DEVELOPER.md]: Something I added

**Comment by fgourichon** [comment]
@cesar-tonnoir thank you for the remarks: ready for review

**Comment by fgourichon** [comment]
@cesar-tonnoir updated with un-existing tag management

**Pull request merged**